querying of named objects in windows across terminal services sessions.  even mutexes.

scottlundgren opened the pull request:
objects such as mutexes, events, semaphores, and other transient
objects.

DESCRIPTION:
Enumerates _all_ named objects in the "BaseNamedObjects" directory within the Windows object directory namespace, across _all_ windows terminal services sessions.   There are ~50 object types, including well-known types such as mutexes (known as mutants by the windows kernel), events, timers, semaphors, ports (for LPC and ALPC), sections (what back memory mappings), and symbolic links.

The implementation consists of one net new virtual table, basenamedobjects, as described in specs/basenamedobjects.table.  The core implementation of that table is in osquery/tables/system/windows/objects.cpp.  The reason for naming the file "objects.cpp" is that future tables (see "futures" below) could be easily implemented in this same module.  A simple C++ wrapper class for Windows HANDLEs was created and implemented in kobjhandle.*, which I put in the osquery/core/windows directory.  kobjhandle stands for Kernel Object Handle.  Also, because working with the Windows object directory and symbolic link object type is entirely unsupported using the win32 api, I added osquery/core/windows/ntapi.h to provide definitions of various windows native api types and functions.

LIMITATONS:
The windows object namespace is hierarchal and allows for arbitrary depth.  This table _only_ exposes the "basenamedobjects" directory across each sessions.  This has high value because this is the default location for all objects created by the win32 api (CreateMutex, CreateEvent, CreateWaitableTimer, CreateSemaphore, etc.).  Other directories are not supported.

THINGS I'VE DONE SO FAR:
I did an initial implementation of the _functionality_ at https://github.com/scottlundgren/objects (no osquery integration).  I then forked osquery at https://github.com/scottlundgren/osquery and pushed the 2754 branch to show support within osquery.  I have done _informal_ code reviews with @zacbrown with primary focus on switching coding style from a procedural, C-inspired style to a (hopefully!) more appropriate C++ style with heavy use of STL.  I have run clang-format against all .h and .cpp files.  I have tested manually on Windows 10 and Windows 7.

THINGS I HAVE NOT DONE:
I have not run any of the automatic testing against this implemenation.  I have not run against any other OS versions other than x64 Windows 10 and Windows 7.  I have not done any "serious" memory, HANDLE, or other resource leak testing.

FUTURES:
There are two obvious follow-up tables to this work.

#4546 - rather than limit to just the basenamedobjects directory, extend to the entire windows object directory namespace.  note that there are benefits to the approach taken in this issue (#2754 - enumerating only basenamedobjects) for two reasons:  first, it is much simpler to execute across terminal services sessions and two: most "threat-hunting" queries will target the basenamedobjects directory, so the simplicity of the query will make it more accessible to a broader audience.

#4403 - enumerate all open HANDLEs across all processes.  Most of these HANDLEs will be to objects in the object namespace, but some will be to objects in other namespaces (notably, filesystems and registry).

akindyakov commented:
ok to test

muffins commented on osquery/core/windows/ntapi.h:
```diff
@@ -0,0 +1,95 @@
+ *
+ *  The purpose of this header is to provide a common place to define Windows
+ *  native api constructs for use across Windows-specific tables
+ *
```
nit: remove the extra `*` here.

muffins commented on osquery/core/windows/kobjhandle.cpp:
```diff
@@ -0,0 +1,123 @@
+
+#define _WIN32_DCOM
+
+#include "Kobjhandle.h"
```
Not sure we need the `k` in the header and impl file/class names? Why not just `objhandle` or `objecthandle`?

muffins replied:
Looks like once you have this as a relative include path the build should at least get beyond it's current error.

scottlundgren replied:
@muffins the reason for the k (for kernel) was an attempt to limit the scope of the class to _just_ kernel objects (as compared to non-kernel objects such as services (SC_HANDLE) etc.  I have _zero_ issue with changing it, just let me know.

the full path on includes is done.  it's building locally without the full path, so I guess I don't understand why need the full path, but no problem obviously changing.

zacbrown replied:
The biggest observation I'd make about that is not all HANDLE "invalid values" are created equally. Some use NULL, some use INVALID_HANDLE. In practice, I believe they're the same but that's not a good thing to rely on :).

scottlundgren replied:
@zacbrown for sure.  INVALID_HANDLE_VALUE is (unsigned long)-1.  NULL is of course 0.

INVALID_HANDLE_VALUE is returned for all fs objects (files, sockets, pipes).  NULL for evertyhig else.

Another key difference is the use of CloseHandle.  For example, even though a HDESKTOP is, in fact, a kernel HANDLE, CloseDesktop() provides some custom logic specific to desktops.  If CloseHandle() is used on a handle to a desktop, it raises an exception.  There are a series of other cases, including some in reverse.  For example, a HKEY is a kernel handle except when it isn't - which includes remote registry psuedo-handles.  If the HKEY references a HANDLE opened to the remote registry service, even if the remeote registry service is on the local windows instance, then of course the HKEY provides a reference to an object on the remote endpoint.

Net it out - absolutely get it.  My hope was that this class could provide the basis of future abstractions.

muffins replied:
@scottlundgren that's fine, I'm definitely down to leave the class, and re: the `k` part, I'd say what about changing the file name to `kernel_object_handles.cpp`? For something like a helper/utility class I think it'd be better to have naming be a bit more explicit, that being said I'd say let's stick with the class, you have compelling points :)

scottlundgren replied:
renamed just handle.[h|cpp] and renamed the class Handle.  Let's see how that looks to someone else's eyes, figured after all my yakking about HKEYs that it would actually be helfpul to abstract away HANDLEs taht aren't really kernel handles so anyone else doesn't have to think about it.

muffins commented:
One more quick note(read nit ^.^) - On casing, we've typically used `camlCase` in the Windows code base, which is a relic of the initial port, but you've got some `snake_case` floating around, which is reflective of the posix side of our code. At some point I think we'll bring all of the windows code over to `snake_case`, but for now I'd say let's continue to have all Windows code use `camlCase` just for consistency.

scottlundgren commented on osquery/core/windows/kobjhandle.cpp:
```diff
@@ -0,0 +1,123 @@
+}
+
+void KObjHandle::close() {
+  if (NULL != _h) {
```
will do

scottlundgren replied:
done

scottlundgren commented on osquery/core/windows/kobjhandle.cpp:
```diff
@@ -0,0 +1,123 @@
+}
+
+bool KObjHandle::valid() {
+  if (NULL != _h) {
```
will do

scottlundgren replied:
done

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+
+// helper to convert a std::wstring to an integer
+// std::stoi can throw, and std::stol doesn't support std::wstring
+static int safe_wstr_to_int(std::wstring str) {
```
@muffins I didn't find the safeStrTol - now I know, thanks, easy swap

scottlundgren replied:
as described in previous comment, safeStrTol does not support std::wstring, so I added a net new helper in conversions.h.

scottlundgren commented on osquery/core/windows/kobjhandle.cpp:
```diff
@@ -0,0 +1,123 @@
+  }
+
+  // look up address of NtOpenSymbolicLinkObject, exported from ntdll
+  NTOPENSYMBOLICLINKOBJECT NtOpenSymbolicLinkObject =
```
the camel case vs snake case - I actually started with camel but switched to snake at @zacbrown 's suggestion.  think he was thinking longer term was to go snake.   more than happy to switch back

muffins replied:
Good call. Let's go ahead and stick with `snake_case`ing then. Might as well start the revolution somewhere :)

scottlundgren replied:
I made an effort to snake_case _variables_.  What I have not done is snake_case function names.  Happy to do whatever on this front.

muffins replied:
@scottlundgren Nah I think we prefer `camlCase` on function names and `snake_case` for local variables. It was more the upper-case `N`, but considering these are reflected function loads I'd also be fine with leaving the naming scheme as you have it to differentiate that these are formal API functions.

scottlundgren replied:
done on both

scottlundgren commented on osquery/core/windows/kobjhandle.cpp:
```diff
@@ -0,0 +1,123 @@
+}
+
+// open a Windows object directory object with DIRECTORY_QUERY
+bool KObjHandle::openDirObj(std::wstring strName) {
```
see previous comment on this subject, happy to switch back to camel

scottlundgren replied:
looks like we are boldly heading into snake_case across osquery code base, even windows.  so leaving as is.  happy to change back if we revert course.

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+//    https://randomsourcecode.wordpress.com/2015/03/14/enumerating-deviceobjects-from-user-mode/
+//    https://msdn.microsoft.com/en-us/library/bb470238(v=vs.85).aspx
+//
+std::vector<obj_name_type_pair> EnumerateObjectNamespace(std::wstring directory) {
```
np

scottlundgren replied:
sticking with snake_case per previous comment

scottlundgren replied:
done

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+  }
+
+  // iterator index is incremented by NtQueryDirectoryObject
+  for (DWORD index = 0;;) {
```
no problem on both

scottlundgren replied:
done on both

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+typedef std::pair<std::wstring, std::wstring> obj_name_type_pair;
+
+// helper to convert a std::wstring to an integer
+// std::stoi can throw, and std::stol doesn't support std::wstring
```
@muffins pointed out that there is another helper already that accomplishes this.

scottlundgren replied:
@muffins @zacbrown the helpers in conversions.h do NOT support std::wstring.  for reasons I documented in the code, I believe I'm compelled to support std::wstring. here's why:

to support object enumeration, I *need* to take output parameters in the form of strings _from_ windows apis and pass them _to_ other windows apis, again as strings, as input parameters.  because windows uses utf-16 strings internally, it is possible that object names can be outside any 8 bit character set and so therefore the only reasonable 8 bit encoding that will work is UTF-8.  That said, I'd be compelled to convert UTF-8 _to_ UTF-16, pass to windows, then convert output back to UTF-8 to use std::string consistently, then re-convert to UTF-16 prior to invoking next Windows API, convert output back to UTF-8, etc.  To net that all out, it seems to make sense to use std::wstring _internaly to table implementation module (objects.cpp)_ and convert to utf-8 (std::string) when returning back to the rest of osquery.

That's a long-winded way of saying that I believe we need std::swstring support in the helper functions.

note that I chose to maintain camel case in the helper in conversions.h, which is NOT consistent with the rest of conversions.h.  I'm happy to convert, just not sure how you guys want to proceed.

note also that I chose not to add a radix or base support for the helper.  Debatalbe choice.

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+    //  get the name and type of the index'th object in the directory
+    NTSTATUS ntStatus = NtQueryDirectoryObject(kdo.getAsHandle(),
+                                               pObjDirInfo,
+                                               sizeof(rgDirObjInfoBuffer),
```
got it

scottlundgren replied:
done

scottlundgren commented:
@zacbrown @muffins - thank you for the feedback.  I will incorporate most all of the changes and push the updates.

Regarding the two bigger points:
* I'm afriad I don't understand the reflection comment @muffins .  can you point me to what you mean so I can understand pls?

* regarding the helper class, I am open to whatever but would make the case for a wrapper class.  I think the (very thin) initial impl will be much bigger over time.  I can make a pass at that in a mod by showing one way that might be possible.

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+
+  // iterator index is incremented by NtQueryDirectoryObject
+  for (DWORD index = 0;;) {
+    BYTE rgDirObjInfoBuffer[1024 * 8] = {0};
```
done

scottlundgren commented:
camelCase vs snake_case discussion resolved in #windows slack channel.  decision out of that conversation was to begin the slow painstaking conversion _from_ camelCase _to_ snake_case in the windows-specific areas with this PR.

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+//    https://randomsourcecode.wordpress.com/2015/03/14/enumerating-deviceobjects-from-user-mode/
+//    https://msdn.microsoft.com/en-us/library/bb470238(v=vs.85).aspx
+//
+std::vector<obj_name_type_pair> EnumerateObjectNamespace(std::wstring directory) {
```
done

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+// "\Sessions\BNOLINKS\<sessionnum>"
+//
+std::vector<obj_name_type_pair>
+EnumerateBaseNamedObjectsLinks(std::wstring session_num,
```
done

muffins replied:
nit: lower on the function name here also, further let's change this to possibly be `Status enumerateBaseNamedObjectsLinks(std::wstring session_num, std::wstring object_type, std::vector<obj_name_type_pai> results) {` to avoid returning a potentially large amount of data.

muffins replied:
nit: lower on the function name here also, further let's change this to possibly be `Status enumerateBaseNamedObjectsLinks(std::wstring session_num, std::wstring object_type, std::vector<obj_name_type_pai> results) {` to avoid returning a potentially large amount of data.

scottlundgren commented on osquery/core/windows/kobjhandle.cpp:
```diff
@@ -0,0 +1,123 @@
+#include "Kobjhandle.h"
+#include "ntapi.h"
+
+#include <Windows.h>
```
done

scottlundgren commented on osquery/core/windows/kobjhandle.cpp:
```diff
@@ -0,0 +1,123 @@
+  }
+}
+
+// open a Windows symbolic link by name with SYMBOLIC_LINK_QUERY
```
done

scottlundgren commented on osquery/core/windows/ntapi.h:
```diff
@@ -0,0 +1,95 @@
+
+#include <osquery/core.h>
+#include <osquery/tables.h>
+#include <windows.h>
```
done

scottlundgren commented on osquery/core/windows/ntapi.h:
```diff
@@ -0,0 +1,95 @@
+                                        PULONG ReturnLength);
+
+typedef NTSTATUS(WINAPI* ZwQuerySystemInformation)(
+    INT /*SYSTEM_INFORMATION_CLASS*/ SystemInformationClass,
```
replaced with proper typedef

scottlundgren commented on osquery/core/windows/kobjhandle.cpp:
```diff
@@ -0,0 +1,123 @@
+}
+
+// open a Windows symbolic link by name with SYMBOLIC_LINK_QUERY
+bool KObjHandle::openSymLinkObj(std::wstring strName) {
```
done

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+  // iterator index is incremented by NtQueryDirectoryObject
+  for (DWORD index = 0;;) {
+    BYTE rgDirObjInfoBuffer[1024 * 8] = {0};
+    POBJDIR_INFORMATION pObjDirInfo = (POBJDIR_INFORMATION)rgDirObjInfoBuffer;
```
ended up with reinterpret_cast here.  if I'm an idiot, let me know and I'll figure this out.

muffins replied:
@scottlundgren you might commonly have to `reinterpret_cast`. We try to push back on those if at all possible, and just really ensure that folks have made attempts to leverage `static_cast` if they can, but Windows doesn't usually let us get away with that :\

scottlundgren replied:
np :)

done

scottlundgren commented on specs/windows/basenamedobjects.table:
```diff
@@ -0,0 +1,12 @@
+table_name("basenamedobjects")
```
this is a super interesting topic, because it ties together the three logically related tables in this general area.  I'll put together a doc.

muffins commented on osquery/core/windows/kobjhandle.cpp:
```diff
@@ -0,0 +1,123 @@
+  close();
+}
+
+HANDLE KObjHandle::getAsHandle() {
```
This is a neat idea. I think it might make the usage more elegant for sure, @scottlundgren if you get a chance I'd say land something like this, but also I'm fine if we just stick with a `get`er method.

muffins commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+
+  // at this point we have SymbolicLink with a name matching a terminal services
+  // session id.  now build the fully qualified object path
+  std::wstring qualifiedpath = L"\\Sessions\\BNOLINKS\\" + session_num;
```
Seems like some of these strings should be `const` at the global scope, might allow for some variable re-use and make constructing things a bit easier?

muffins replied:
Let's also make this `const auto`.

muffins commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+  QueryData results;
+
+  // enumerate the base named objects in each terminal services session
+  auto sessions = EnumerateObjectNamespace(L"\\Sessions\\BNOLINKS");
```
Can swap this out once you have the global `const` string.

scottlundgren commented on osquery/core/windows/ntapi.h:
```diff
@@ -0,0 +1,95 @@
+#define NTSTATUS ULONG
+#define STATUS_SUCCESS 0L
+
+#define OBJ_CASE_INSENSITIVE 64L
```
happy to do that.  what I'm doing with the #defines here is mimic'ing the nt ddk.  I have zero issue updating, just want to confirm before I do @muffins

muffins replied:
@scottlundgren I'm down to leave it, especially considering this is the NT API for osquery

scottlundgren commented:
I _believe_ I've addressed everything except these three points:

operator HANDLE() suggested by @zacbrown - this is forthcoming
static const unsigned long rather than #define for Windows native API constants - see open q for @muffins 
I have not yet created the document around proposed schema and therefore have not (yet) changed the table name.

major changes out of this push were introduction of osquery::Status as return type, incorporated hopefully all the feedback other than what was provided.

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+#include <Windows.h>
+
+#include <osquery/core.h>
+#include <osquery/core/windows/kobjhandle.h>
```
done

scottlundgren commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,214 @@
+//    https://randomsourcecode.wordpress.com/2015/03/14/enumerating-deviceobjects-from-user-mode/
+//    https://msdn.microsoft.com/en-us/library/bb470238(v=vs.85).aspx
+//
+std::vector<obj_name_type_pair> EnumerateObjectNamespace(std::wstring directory) {
```
done.  now returning osqeruy::Status

scottlundgren commented:
I have updated the tablename per last comments from @muffins and also included a larger document describing an overall strategy at https://docs.google.com/document/d/1BadhBDXMJvFiSq7_raSrGGh54KSq1jw91ANYDSw2xzc/edit#

The tablename 'winbno' is proposed for #2754 and 'winobj' for #4546

muffins commented on osquery/core/windows/handle.cpp:
```diff
@@ -0,0 +1,118 @@
+  }
+}
+
+bool Handle::openSymLinkObj(const std::wstring& strName) {
```
Looks like this should be `Status Handle::openSymLinkObj(const std::wstring& strName)` as you that's how you have it declared and seems to be how you're expecting to use it in your table code. Perhaps this could `return Status(1, "Open symbolic link to object failed with " + std::to_string(GetLastError()));` or `return Status();` otherwise?

scottlundgren replied:
@muffins ok got it

scottlundgren replied:
done

muffins commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,229 @@
+  Handle slo;
+  auto status = slo.openSymLinkObj(qualifiedpath);
+  if (!status.ok()) {
+    // log?
```
Yeah, once you fix up the function definition, you could maybe `VLOG(1) << status.getMessage()`?

scottlundgren replied:
@muffins ok will do!

scottlundgren replied:
done

muffins commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,229 @@
+  // enumerate the base named objects in each terminal services session
+  //
+  std::vector<obj_name_type_pair> sessions;
+  auto status = enumerateObjectNamespace(L"\\Sessions\\BNOLINKS", sessions);
```
Looks like we use `L"\\Sessions\\BNOLINKS"` more than once, maybe make it a const global? `const std::wstring kBnoLinks { L"\\Sessions\\BNOLINKS" };`?

scottlundgren replied:
@muffins ok great will do

scottlundgren replied:
done

muffins commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,229 @@
+  std::vector<obj_name_type_pair> sessions;
+  auto status = enumerateObjectNamespace(L"\\Sessions\\BNOLINKS", sessions);
+  if (!status.ok()) {
+    // log warning?
```
Same here, I think we've discussed making verbosity in tables be a `VLOG(1)` or something that might not generate loads of noise in the event the query isn't written well, /cc @fmanco @mkareta @obelisk to correct me if I'm wrong on this.

scottlundgren replied:
done

akindyakov commented on osquery/core/conversions.h:
```diff
@@ -156,6 +156,20 @@ inline Status safeStrtol(const std::string& rep, size_t base, long int& out) {
+    return 0;
+  }
+}
+
```
There is a new way to convert string to numbers in `conversions.h`: `tryTo<in>(...)`. Check it out :)

scottlundgren replied:
@akindyakov sure thing!

akindyakov commented on osquery/core/windows/ntapi.h:
```diff
@@ -0,0 +1,108 @@
+  SystemTimeOfDayInformation,
+  SystemPathInformation,
+  SystemProcessInformation
+} SYSTEM_INFORMATION_CLASS;
```
As far as I know you could write is in more C++ way, like:
```c++
enum  SYSTEM_INFORMATION_CLASS {
  SystemProcessorInformation,
  SystemPerformanceInformation,
  SystemTimeOfDayInformation,
  SystemPathInformation,
  SystemProcessInformation
};
```
right?

scottlundgren replied:
@akindyakov the reasoning behind how "ntapi.h" is constructed was to mimic the Microsoft defintions, including the style.  The idea was to be consistent with MS-style declarations in this MS-specific header.  any thoughts on this?

muffins replied:
My 2-cents -- I'd say let's leave this for consistency, as it's a standard API that one can find in many MS API interfaces. That being said I'd be curious of @akindyakov's thoughts on it

akindyakov replied:
There is no big difference, so we can keep it as-is, no problem.

scottlundgren replied:
@akindyakov appreciated, thanks!

akindyakov commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,229 @@
+// arbitrary buffer size to be used when querying for a single
+// object-name--object-type-name pair with NtQueryDirectoryObject
+//
+static const unsigned long obj_buf_size = 8 * 1024;
```
`osquery` has a bit different way to name global consts:
```c++
static const unsigned long kObjBufSize = 8 * 1024;
```

scottlundgren replied:
@akindyakov @muffins got it, will address.

scottlundgren replied:
done

akindyakov commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,229 @@
+// converted to UTF-8 encoded std::strings when communicating with the
+// rest of osquery
+//
+typedef std::pair<std::wstring, std::wstring> obj_name_type_pair;
```
Prefer `using` over `typedef` please.

scottlundgren replied:
@akindyakov ok sure thing

scottlundgren replied:
using obj_name_type_pair = std::pair<std::wstring, std::wstring>;

@akindyakov done

akindyakov commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,229 @@
+  // look up NtQuerySymbolicLinkObject as exported from ntdll
+  //
+  auto NtQuerySymbolicLinkObject =
+      (NTQUERYSYMBOLICLINKOBJECT)GetProcAddress(GetModuleHandleA("ntdll"),
```
Prefer C++ casts over unsafe C way, [some explanation](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es49-if-you-must-use-a-cast-use-a-named-cast).

scottlundgren replied:
@akindyakov no problem

scottlundgren replied:
both are using reinterpret_cast now

scottlundgren replied:
done

akindyakov commented on osquery/tables/system/windows/objects.cpp:
```diff
@@ -0,0 +1,229 @@
+  auto NtQuerySymbolicLinkObject =
+      (NTQUERYSYMBOLICLINKOBJECT)GetProcAddress(GetModuleHandleA("ntdll"),
+                                                "NtQuerySymbolicLinkObject");
+  if (NULL == NtQuerySymbolicLinkObject) {
```
It is much better to use `nullptr` instead of `NULL`, [some explanation](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es47-use-nullptr-rather-than-0-or-null)

scottlundgren replied:
no problem

scottlundgren replied:
done

muffins commented:
ok to test
